File: server/core/src/main/java/dev/slimevr/tracking/processor/skeleton/TapDetectionManager.java
package dev.slimevr.tracking.processor.skeleton;


import dev.slimevr.config.TapDetectionConfig;
import dev.slimevr.reset.ResetHandler;
import dev.slimevr.setup.TapSetupHandler;
import dev.slimevr.tracking.processor.HumanPoseManager;
import dev.slimevr.tracking.trackers.Tracker;
import solarxr_protocol.rpc.ResetType;

import java.util.ArrayList;
import java.util.List;


/**
 * Handles tap detection for reset
 */
public class TapDetectionManager {
	private static final String resetSourceName = "TapDetection";
	private static final int tapsForSetupMode = 2;

	// server and related classes
	private final HumanSkeleton skeleton;
	private HumanPoseManager humanPoseManager;
	private TapDetectionConfig config;

	// tap detectors
	private TapDetection yawResetDetector;
	private TapDetection fullResetDetector;
	private TapDetection mountingResetDetector;

	private ArrayList<TapDetection> tapDetectors;

	// number of taps to detect
	private int yawResetTaps = 2;
	private int fullResetTaps = 3;
	private int mountingResetTaps = 3;

	// delay
	private static final float NS_CONVERTER = 1.0e9f;
	private float fullResetDelayNs = 0.20f * NS_CONVERTER;
	private float yawResetDelayNs = 1.00f * NS_CONVERTER;
	private float mountingResetDelayNs = 1.00f * NS_CONVERTER;

	// feedback
	private boolean yawResetAllowPlaySound = true;
	private boolean fullResetAllowPlaySound = true;
	private boolean mountingResetAllowPlaySound = true;

	private ResetHandler resetHandler;
	private TapSetupHandler tapSetupHandler;

	public TapDetectionManager(HumanSkeleton skeleton) {
		this.skeleton = skeleton;
	}

	public TapDetectionManager(
		HumanSkeleton skeleton,
		HumanPoseManager humanPoseManager,
		TapDetectionConfig config,
		ResetHandler resetHandler,
		TapSetupHandler tapSetupHandler,
		List<Tracker> trackers
	) {
		this.skeleton = skeleton;
		this.humanPoseManager = humanPoseManager;
		this.config = config;
		this.resetHandler = resetHandler;
		this.tapSetupHandler = tapSetupHandler;

		updateConfig(trackers);
	}

	public void updateConfig(List<Tracker> trackers) {
		// check the skeleton for new trackers
		yawResetDetector = new TapDetection(skeleton, getTrackerToWatchYawReset());
		fullResetDetector = new TapDetection(skeleton, getTrackerToWatchFullReset());
		mountingResetDetector = new TapDetection(skeleton, getTrackerToWatchMountingReset());

		if (trackers != null) {
			tapDetectors = new ArrayList<>();
			for (Tracker tracker : trackers) {
				TapDetection tapDetector = new TapDetection(skeleton, tracker);
				tapDetector.setEnabled(true);
				tapDetectors.add(tapDetector);
			}
		}

		if (this.config == null) {
			return;
		}

		this.yawResetDelayNs = config.getYawResetDelay() * NS_CONVERTER;
		this.fullResetDelayNs = config.getFullResetDelay() * NS_CONVERTER;
		this.mountingResetDelayNs = config.getMountingResetDelay() * NS_CONVERTER;
		yawResetDetector.setEnabled(config.getYawResetEnabled());
		fullResetDetector.setEnabled(config.getFullResetEnabled());
		mountingResetDetector.setEnabled(config.getMountingResetEnabled());
		yawResetTaps = config.getYawResetTaps();
		fullResetTaps = config.getFullResetTaps();
		mountingResetTaps = config.getMountingResetTaps();
		yawResetDetector.setMaxTaps(yawResetTaps);
		fullResetDetector.setMaxTaps(fullResetTaps);
		mountingResetDetector.setMaxTaps(mountingResetTaps);
		yawResetDetector
			.setNumberTrackersOverThreshold(
				config.getNumberTrackersOverThreshold()
			);
		fullResetDetector
			.setNumberTrackersOverThreshold(
				config.getNumberTrackersOverThreshold()
			);
		mountingResetDetector
			.setNumberTrackersOverThreshold(
				config.getNumberTrackersOverThreshold()
			);
	}

	public void update() {
		if (
			yawResetDetector == null
				|| fullResetDetector == null
				|| mountingResetDetector == null
				|| tapDetectors == null
				|| config == null
		)
			return;

		// if setup mode is enabled, update the tap detectors for each tracker
		if (config.getSetupMode()) {
			for (TapDetection tapDetector : tapDetectors) {
				tapDetector.update();

				if (tapDetector.getTaps() >= tapsForSetupMode) {
					tapSetupHandler.sendTap(tapDetector.getTracker());
					tapDetector.resetDetector();
				}
			}
		} else {
			// update the tap detectors
			yawResetDetector.update();
			fullResetDetector.update();
			mountingResetDetector.update();

			// check if any tap detectors have detected taps
			checkYawReset();
			checkFullReset();
			checkMountingReset();
		}
	}

	private void checkYawReset() {
		boolean tapped = (yawResetTaps <= yawResetDetector.getTaps());

		if (tapped && yawResetAllowPlaySound) {
			this.resetHandler.sendStarted(ResetType.Yaw);
			yawResetAllowPlaySound = false;
		}

		if (
			tapped && System.nanoTime() - yawResetDetector.getDetectionTime() > yawResetDelayNs
		) {
			if (humanPoseManager != null)
				humanPoseManager.resetTrackersYaw(resetSourceName);
			else
				skeleton.resetTrackersYaw(resetSourceName);

			yawResetDetector.resetDetector();
			yawResetAllowPlaySound = true;
		}
	}

	private void checkFullReset() {
		boolean tapped = (fullResetTaps <= fullResetDetector.getTaps());

		if (tapped && fullResetAllowPlaySound) {
			this.resetHandler.sendStarted(ResetType.Full);
			fullResetAllowPlaySound = false;
		}

		if (
			tapped && System.nanoTime() - fullResetDetector.getDetectionTime() > fullResetDelayNs
		) {
			if (humanPoseManager != null)
				humanPoseManager.resetTrackersFull(resetSourceName);
			else
				skeleton.resetTrackersFull(resetSourceName);

			fullResetDetector.resetDetector();
			fullResetAllowPlaySound = true;
		}
	}

	private void checkMountingReset() {
		boolean tapped = (mountingResetTaps <= mountingResetDetector.getTaps());

		if (tapped && mountingResetAllowPlaySound) {
			this.resetHandler.sendStarted(ResetType.Mounting);
			mountingResetAllowPlaySound = false;
		}

		if (
			tapped
				&& System.nanoTime() - mountingResetDetector.getDetectionTime()
					> mountingResetDelayNs
		) {
			skeleton.resetTrackersMounting(resetSourceName);
			mountingResetDetector.resetDetector();
			mountingResetAllowPlaySound = true;
		}
	}


	// returns either the chest tracker, hip tracker, or waist tracker depending
	// on which one is available
	// if none are available, returns null
	private Tracker getTrackerToWatchYawReset() {
		if (skeleton.getUpperChestTracker() != null)
			return skeleton.getUpperChestTracker();
		else if (skeleton.getChestTracker() != null)
			return skeleton.getChestTracker();
		else if (skeleton.getHipTracker() != null)
			return skeleton.getHipTracker();
		else if (skeleton.getWaistTracker() != null)
			return skeleton.getWaistTracker();
		else
			return null;
	}

	private Tracker getTrackerToWatchFullReset() {
		if (skeleton.getLeftUpperLegTracker() != null)
			return skeleton.getLeftUpperLegTracker();
		else if (skeleton.getLeftLowerLegTracker() != null)
			return skeleton.getLeftLowerLegTracker();
		return null;
	}

	private Tracker getTrackerToWatchMountingReset() {
		if (skeleton.getRightUpperLegTracker() != null)
			return skeleton.getRightUpperLegTracker();
		else if (skeleton.getRightLowerLegTracker() != null)
			return skeleton.getRightLowerLegTracker();
		return null;
	}

}

File: server/core/src/main/java/dev/slimevr/tracking/processor/skeleton/TapDetectionManager.kt
package dev.slimevr.tracking.processor.skeleton

import dev.slimevr.VRServer
import dev.slimevr.tracking.processor.HumanPoseManager
import dev.slimevr.tracking.trackers.Tracker
import dev.slimevr.tracking.trackers.TrackerPosition
import dev.slimevr.tracking.trackers.TrackerUtils
import io.eiren.util.logging.LogManager
import java.util.concurrent.CopyOnWriteArrayList

class TapDetectionManager(
	val server: VRServer,
	val skeleton: HumanSkeleton,
	val humanPoseManager: HumanPoseManager,
) {
	private var tapDetectors: MutableList<TapDetection> = CopyOnWriteArrayList()
	var yawResetDetector: TapDetection? = null
	var fullResetDetector: TapDetection? = null
	var mountingResetDetector: TapDetection? = null

	var config = server.configManager.vrConfig.tapDetection

	init {
		refresh()
	}

	fun registerSingleTapDetectors() {
		for (tracker in server.allTrackers) {
			tapDetectors.add(
				TapDetection(skeleton, tracker, config.numberTrackersOverThreshold, 2) {
					server.tapSetupHandler.sendTap(tracker)
				},
			)
		}
	}

	fun registerResetsDetectors() {
		val yawTracker = yawResetTracker
		yawResetDetector = if (yawTracker != null && config.yawResetEnabled) {
			TapDetection(skeleton, yawTracker, config.numberTrackersOverThreshold, config.yawResetTaps) {
				server.scheduleResetTrackersYaw(RESET_SOURCE_NAME, (config.yawResetDelay * 1000).toLong())
			}
		} else {
			null
		}

		val fullTracker = fullResetTracker
		fullResetDetector = if (fullTracker != null && config.fullResetEnabled) {
			TapDetection(skeleton, fullTracker, config.numberTrackersOverThreshold, config.fullResetTaps) {
				server.scheduleResetTrackersFull(RESET_SOURCE_NAME, (config.fullResetDelay * 1000).toLong())
			}
		} else {
			null
		}

		val mountingTracker = mountingResetTracker
		mountingResetDetector = if (mountingTracker != null && config.mountingResetEnabled) {
			TapDetection(skeleton, mountingTracker, config.numberTrackersOverThreshold, config.mountingResetTaps) {
				server.scheduleResetTrackersMounting(RESET_SOURCE_NAME, (config.mountingResetDelay * 1000).toLong())
			}
		} else {
			null
		}
	}

	/**
	 * Called when the list of available trackers gets updated
	 * or when the tap settings get changed
	 * it re-create the tap detectors according to the configs and available trackers
	 */
	fun refresh() {
		tapDetectors.clear()
		registerSingleTapDetectors()
		registerResetsDetectors()
		LogManager.info(yawResetTracker.toString())
	}

	fun update() {
		// We disable the resets detectors during the assignment phase so you cant
		// trigger a reset while assigning
		if (config.setupMode) {
			for (detector in tapDetectors) {
				detector.update()
			}
		} else {
			yawResetDetector?.update()
			fullResetDetector?.update()
			mountingResetDetector?.update()
		}
	}

	private val mountingResetTracker: Tracker?
		get() {
			return arrayOf(
			TrackerUtils.getTrackerForSkeleton(server.allTrackers, config.mountingResetTracker),
				if (config.yawResetTracker !== TrackerPosition.RIGHT_UPPER_LEG && config.fullResetTracker !== TrackerPosition.RIGHT_UPPER_LEG) skeleton.rightUpperLegTracker else null,
				if (config.yawResetTracker !== TrackerPosition.RIGHT_LOWER_LEG && config.fullResetTracker !== TrackerPosition.RIGHT_LOWER_LEG) skeleton.rightLowerLegTracker else null
			).firstNotNullOfOrNull { it }
		}

	private val fullResetTracker: Tracker?
		get() {
			return arrayOf(
				TrackerUtils.getTrackerForSkeleton(server.allTrackers, config.fullResetTracker),
				if (config.yawResetTracker !== TrackerPosition.LEFT_UPPER_LEG && config.mountingResetTracker !== TrackerPosition.LEFT_UPPER_LEG) skeleton.leftUpperLegTracker else null,
				if (config.yawResetTracker !== TrackerPosition.LEFT_LOWER_LEG && config.mountingResetTracker !== TrackerPosition.LEFT_LOWER_LEG) skeleton.leftLowerLegTracker else null
			).firstNotNullOfOrNull { it }
		}

	private val yawResetTracker: Tracker?
		get() {
			return arrayOf(
				TrackerUtils.getTrackerForSkeleton(server.allTrackers, config.yawResetTracker),
				if (config.fullResetTracker !== TrackerPosition.UPPER_CHEST && config.mountingResetTracker !== TrackerPosition.UPPER_CHEST) skeleton.upperChestTracker else null,
				if (config.fullResetTracker !== TrackerPosition.CHEST && config.mountingResetTracker !== TrackerPosition.CHEST) skeleton.chestTracker else null,
				if (config.fullResetTracker !== TrackerPosition.HIP && config.mountingResetTracker !== TrackerPosition.HIP) skeleton.hipTracker else null,
				if (config.fullResetTracker !== TrackerPosition.WAIST && config.mountingResetTracker !== TrackerPosition.WAIST) skeleton.waistTracker else null
			).firstNotNullOfOrNull { it }
		}

	companion object {
		const val RESET_SOURCE_NAME: String = "TapDetection"
	}
}
